File: hw7/141044034/BigramMap.cpp
#include <vector>
#include <string>
#include <iostream>
#include <cstdlib>
#include <fstream>
#include <map>
#include "Bigram.h"
#include "BigramMap.h"

using namespace std;
template <class T>
void BigramMap<T>::readFile(const char* fileName)throw(myexception)
{
    ifstream dataFile(fileName);
    if(dataFile.is_open() == false)
    {
        throw(exc);
    }
    //bool a;
     for(int i=0 ; !dataFile.eof();i++){
        if(!(dataFile >> data[i]))
            throw(exc);
        /*if(a == false)
            throw()*/
    }    

    for(int i=0;i < data.size()-1;i++){

        bigrams[i].first=data[i];
        bigrams[i].second=data[i+1];
    
    }
}

template <class T>
int BigramMap<T>::numGrams()
{
    return bigrams.size();
}

template <class T>
int BigramMap<T>::numOfGrams(T p1 , T p2)
{
    int count=0;
    pair<T ,T> check;
    check.first =p1;
    check.second=p2;

    for(int i=0;i<bigrams.size();i++)
    {

        if(check == bigrams[i])
            ++count;
    }

    return count;
}


template <class T>
pair<T,T> BigramMap<T>::maxGrams()
{
    int tempIndex=-1;
    int tempOccurence=0;
    for(int i =0;i<bigrams.size();i++)
    {

        if(numOfGrams(bigrams[i].first,bigrams[i].second) > tempOccurence )
        {
            tempOccurence = numOfGrams(bigrams[i].first,bigrams[i].second);
            tempIndex = i ;
        }
    }
    return bigrams[tempIndex];
}


template <class T>
ostream& BigramMap<T>::print(ostream& os) 
{

    int tempIndex=-1;
    int tempOccurence=0;
    int i;
    pair<T,T> tempP;
    map <int , pair<T,T> > tempB; 

    for(i=0;i<bigrams.size();i++){
        tempB[i].first=bigrams[i].first;
        tempB[i].second=bigrams[i].second;
    }


    for(int j=0;j<tempB.size();j++){
        for(int i =j;i<tempB.size();i++)
        {
            if(numOfGrams(tempB[i].first,tempB[i].second) > tempOccurence )
            {
                tempOccurence = numOfGrams(tempB[i].first,tempB[i].second);
                tempIndex = i ;
            }
        }

        tempP.first=tempB[tempIndex].first;
        tempP.second=tempB[tempIndex].second;
        
        tempB[tempIndex].first = tempB[j].first;
        tempB[tempIndex].second = tempB[j].second;
        
        tempB[j].first = tempP.first;
        tempB[j].second = tempP.second;

        tempOccurence=0;
    }

    for(i =0 ; i<bigrams.size();i++)
    {
        os<<"Bigram = " << tempB[i].first << "--"<<tempB[i].second <<endl;
        os<< "Occurence: " << numOfGrams(tempB[i].first,tempB[i].second)<<endl; 

        i+=numOfGrams(tempB[i].first,tempB[i].second)-1;
    }
    return os;

}

template <class T>
BigramMap<T>::BigramMap(int dType){

    dataType=dType;

}

File: hw8/hw8/BigramMap.java

import java.io.*;
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;
 

public class BigramMap<T> implements Bigram<T>{

    private int dType;    
    private Map<Integer,Pair> Bigrams = new HashMap() ;
    Object[] Data ; 
    BigramMap(int dataType){
        dType =dataType;
    }
    
    @Override
    public void readFile(String fileName)throws IOException
    {
        String dataLine;
                
        try(InputStream fileIS = new FileInputStream(fileName);
        InputStreamReader ISReader = new InputStreamReader(fileIS);
        BufferedReader BReader = new BufferedReader(ISReader);){
        while((dataLine =BReader.readLine()) !=null)
            Data = dataLine.split(" ");          
        }
  
        for(int i=0;i < Data.length - 1;i++){

            Pair temp = new Pair();
            temp.first = Data[i];
            temp.second = Data[i+1];
            Bigrams.put(i,temp);
        }
    }
     
    @Override
    public int numGrams()
    {
         return Bigrams.size(); 
    }
    
    @Override
    public int numOfGrams(T p1 , T p2)
    {
        int count=0;
        
        for(int i=0;i<Bigrams.size();i++)
        {
            if(p1.toString().equals(Bigrams.get(i).first.toString()) && p2.toString().equals( Bigrams.get(i).second.toString()))
                ++count;
        }
    

    return count;
    }
    @Override
    public String toString() 
{

    int tempIndex=-1;
    int tempOccurence=0;

    Pair tempP=new Pair();
    Pair tempPB = new Pair();
    Map <Integer , Pair > tempB = new HashMap(); 

    tempB.putAll(Bigrams);
      

    for(int j=0;j<tempB.size();j++){
        for(int i =j;i< tempB.size();i++)
        {
            if(numOfGrams((T)tempB.get(i).first , (T)tempB.get(i).second ) > tempOccurence )
            {
                tempOccurence = numOfGrams((T)tempB.get(i).first,(T)tempB.get(i).second);
                tempIndex = i ;
            }
        }
        
        tempP.first=tempB.get(tempIndex).first;
        tempP.second=tempB.get(tempIndex).second;
        
        tempB.get(tempIndex).first=tempB.get(j).first;
        tempB.get(tempIndex).second=tempB.get(j).second;

        tempB.get(j).first=tempP.first;
        tempB.get(j).second=tempP.second;
        tempOccurence=0;
    }

    
    for(int i =0 ; i<Bigrams.size();i++)
    {
        System.out.print("Bigram = " + tempB.get(i).first+ "--"+tempB.get(i).second );
        System.out.print("Occurence: " + numOfGrams((T)tempB.get(i).first,(T)tempB.get(i).second)); 

        i+=numOfGrams((T)tempB.get(i).first,(T)tempB.get(i).second);
    }
    return "";
}
}
